design/volume-snapshot-data-movement/volume-snapshot-data-movement.md

```diff
@@ -0,0 +1,763 @@
+- Reuse the existing Velero generic data path as creatd in [Unified Repository design][1]
+
+## Non-Goals
+
+- The implementation details of block level access will be included in a future design
+- Most of the components are generic, but the Exposer is snapshot type specific or snapshot access specific. The current design covers the implementation details for exposing CSI snapshot to host path access only, for other types or accesses, we may need a separate design
+- The current workflow focuses on snapshot-based data movements. For some application/SaaS level data sources, snapshots may not be taken explicitly. We don’t take them into consideration, though we believe that some workflows or components may still be reusable.
+
+## Architecture of Volume Snapshot Data Movement
+
+## Workflows  
+
+Here are the diagrams that illustrate components and workflows for backup and restore respectively.  
+For backup, we intend to create an extensive architecture for various snapshot types, snapshot accesses and various data accesses. For example, the snapshot specific operations are isolated in Data Mover Plugin and Exposer. In this way, we only need to change the two modules for variations. Likely, the data access details are isolated into uploaders, so different uploaders could be plugged into the workflow seamlessly.  
+
+For restore, we intend to create a generic workflow that could for all backups. This means the restore is backup source independent. Therefore, for example, we can restore a CSI snapshot backup to another cluster with no CSI facilities or a CSI driver the same as the source cluster.  
+We still have the Exposer module for restore and it is to expose the target volume to the data path. Therefore, we still have the flexibility to introduce different ways to expose the target volume.  
+Likely, the data downloading details are isolated in uploaders, so we can still create multiple types of uploaders.  
+
+Below is the backup workflow:  
+![backup-workflow.png](backup-workflow.png)  
+
+Below is the restore workflow:  
+![restore-workflow.png](restore-workflow.png)  
+
+## Components
+Below are the generic components in the data movement workflow:  
+
+**Velero**: Velero controls the backup/restore workflow, it calls BIA/RIA V2 to backup/restore an object that involves data movement, specifically, a PVC or a PV.  
+**BIA/RIA V2**: BIA/RIA V2 are the protocols between Velero and the data mover plugins. They support asynchronized operations so that Velero backup/restore is not marked as completion until the data movement is done and in the meantime, Velero is free to process other backups during the data movement.  
+**Data Mover Plugin (DMP)**: DMP implement BIA/RIA V2 and it invokes the corresponding data mover by creating the DataUpload/DataDownload CRs. DMP is also responsible to take snapshot of the source volume, so it is a snapshot type specific module. For CSI snapshot data movement, the CSI plugin could be extended as a DMP.  
```

Comment:
My understanding is there should be a new plugin created for VBDM, and the current CSI plugin could be a base to create the new VBDM, so there still is only one Velero CSI plugin.

Reply:
We incline to take the first action because V2 is the new standard and it is a good opportunity (modify for data mover) to port the CSI plugin with the new standard. However, if we have some strong requirement against this, it is OK to take the second action technically, so if any, please advice.

Reply:
Let me clarify it in the design as the first action, if we finally want to change it, we come back to change the design.

Reply:
Done for the clarification, check line 60:

> this also means that the CSI plugin will fully implement BIA/RIA V2

Reply:
@Lyndon-Li Keep in mind that with v2 plugins, any plugins that kick off asynchronous processes need to be included in the original backup processing (InProgress stage via backup controller). i.e. if we make CSI plugin v2 and also introduce a returned operationID, any new resources created by this plugin that are added during finalize can't have their own async operations -- however, if the CSI plugin returns an AdditionalItem as a regular backup item to be processed now, then that item may have an async plugin, since it will be processed during the initial InProgress stage, not during Finalize. The issue is that once we hit Finalize, we're done waiting for plugin async operations, so if a finalize item's plugin returns an operationID, it will raise an error.

Reply:
@sseago Thanks for reminding.
This situation could fit to the current data movement case, specifically:
1. After CSI plugin changes to V2, it still works on PVC, VS, VSC, etc., it won't work on DataUpload/DataDownload CR
2. If the data movement is not enabled, nothing is changed, no operationID is returned, VS, VSC are returned as additonalItems as is
3. If the data movement is enabled, an operationID is returned and a DataUpload/DataDownload CR is returned as itemToFinalize. However, DataUpload/DataDownload CR right meets your above statements -- they don't need to go through the regular backup process and the CSI plugin won't handle them on finalize phase, so no further operationID is returned for them during finalize
4. If the data movement is enabled, depend on the discussion[ here](https://github.com/vmware-tanzu/velero/pull/5968#discussion_r1151478358), if we still need to backup VS and VSC, we still return them as regular additionalItems, so they will be processed as is